Feature rename sql objects

M-Patrone opened the pull request:
I’m trying to implement [this feature request](https://github.com/microsoft/azuredatastudio/issues/8511) from the Azure Data  Studio. So, I wanted to add the renaming function to the sqltoolsservice project. To possible enable that feature, I added a service class, which executes the renaming operation. The class “RenameUtils” is my helper class, which validates the parameters. In addition, it creates the SQL-command for renaming objects. In the feature request it is only mentioned to enable renaming tables and columns. If this PR is accepted and the first draft of the feature for the Azure Data Studio is implemented, there is the possibility to extend to database and indexes.  As I am a newbie on this project, I would love to get some feedback. 😊  

Also have a look if I declared the custom error message on the right file. Because I am not 100% sure If I get how the translation works. 

Furthermore, I would love to add my docker file to the wiki, so it would be easier for newbies to execute the integration tests on a SQL server. Is there a specific reason why this would not be wanted? 

Checklist Contribution Guidelines
- [x] Unit- and integrationtests are added and run successfuly
- [x] The 'Release' build and unit tests pass locally 
- [ ] The are two integration tests which didn't run successfully

1. BackupRestoreUrlTests.BackupDatabaseToUrlAndRestoreFromUrlTest() --> I don't have any Azure environment to execute this test
2. ObjectExplorerServiceTests.VerifyAllSqlObjects() -->I didn't check why this isn't working. This test doesn’t work on the main branch either. For that matter, the other test also. Therefore, I didn’t invest time in that. 

Finally, sorry for my bad english 😉

Charles-Gagnon reviewed (commented):
Thanks for the submission! Let me take a look and think about this approach, but it seems like a reasonable enough thing to include.

@kburtram FYI if you wanted to have anyone else on your side look things over or had thoughts about the implementation yourself. 

As for the Docker file, that sounds like a useful thing to have. Is there a reason you wanted to put it in the wiki instead of having it checked into the repo directly? And would this be for development, testing or both?

Charles-Gagnon commented on src/Microsoft.SqlTools.ServiceLayer/Rename/RenameUtils.cs:
```diff
@@ -0,0 +1,79 @@
+            {
+                return String.Format(@"
+                USE [{0}];
+                    EXEC sp_rename @objname = '{1}', @newname = '{2}', @objtype ='{3}';
```
Instead of directly defining the query here I wonder if we should instead be using SMO to handle that for us - that'd save us from duplicating any logic needed. 

@shueybubbles SSMS uses SMO for this, right? Is there any reason you can think of that it wouldn't be a good idea to do that here?

M-Patrone replied:
I did not know that there is an extra project for this (SMO). Nonetheless, I am open if I should include this.   

The only reason I could think of for keeping this would be if at some point in Azure Data Studio it would be necessary to script the renaming operation. Through this you would have the script right away. However, I don't know if the SMO project can possibly do this already.

M-Patrone commented:
@Charles-Gagnon  Thank you very much for your comment 😊. Regarding the Dockerfile: I just thought that this may not have been desired. But I am also for the inclusion of the Docker file. I have this now especially needed for the integration tests. However, I can well imagine that you could also super develop with it. Should I make a separate PR for that?

M-Patrone commented:
> Thanks for the submission! Let me take a look and think about this approach, but it seems like a reasonable enough thing to include.
> 
> @kburtram FYI if you wanted to have anyone else on your side look things over or had thoughts about the implementation yourself.
> 
> As for the Docker file, that sounds like a useful thing to have. Is there a reason you wanted to put it in the wiki instead of having it checked into the repo directly? And would this be for development, testing or both?

@Charles-Gagnon Thank you very much for your comment 😊. Regarding the Dockerfile: I just thought that this may not have been desired. But I am also for the inclusion of the Docker file. I have this now especially needed for the integration tests. However, I can well imagine that you could also super develop with it. Should I make a separate PR for that?

shueybubbles reviewed (changes requested):
Just about every object type in SMO already implements the `Rename` method, you should just call it.
To get a script for the operation instead of running it you can use `Server.ExecuteWithModes`. See https://github.com/microsoft/sqlmanagementobjects/blob/98093604e9ede97aa2a9b066279f0bd9d9557c4f/src/FunctionalTest/Framework/Helpers/TSqlScriptingHelper.cs#L22

M-Patrone commented:
> Just about every object type in SMO already implements the `Rename` method, you should just call it. To get a script for the operation instead of running it you can use `Server.ExecuteWithModes`. See https://github.com/microsoft/sqlmanagementobjects/blob/98093604e9ede97aa2a9b066279f0bd9d9557c4f/src/FunctionalTest/Framework/Helpers/TSqlScriptingHelper.cs#L22

Perfectly. So I will change the logic of that. I think I will open a new PR for that. @Charles-Gagnon and @shueybubbles do you agree with this approach?

Charles-Gagnon commented:
A separate PR for the Dockerfile would be great, thank you!

> Perfectly. So I will change the logic of that. I think I will open a new PR for that. @Charles-Gagnon and @shueybubbles do you agree with this approach?

A new PR for adding rename functionality using SMO? I'd definitely suggest closing this one or converting it to a draft while you work on it (draft might be easier in case you had questions along the way).

shueybubbles commented on src/Microsoft.SqlTools.ServiceLayer/Rename/Requests/ProcessRenameEditRequest.cs:
```diff
@@ -0,0 +1,22 @@
+    /// Property class for Rename Service
+    /// </summary>
+    public class ProcessRenameEditRequestParams : GeneralRequestDetails
+    {
```
the simplest parameter to pass to identify any SMO object is its Urn. I think you could just pass the Urn and the new name.
Use Server.GetSmoObject(urn) to get the object instance.

The only pain point is SMO doesn't appear to define an interface for "Rename" so you will need to have a case statement on the type or use reflection to get the Reflect method to call.

Charles-Gagnon commented:
It looks like your latest merges got messed up - you should merge in the latest from main to your branch and that should fix it. If not you may need to reset your branch to before you brought in the additional commits and then do a merge to bring it up to date.

M-Patrone commented:
> It looks like your latest merges got messed up - you should merge in the latest from main to your branch and that should fix it. If not you may need to reset your branch to before you brought in the additional commits and then do a merge to bring it up to date. (and then reapply any commits you authored yourself)

Sorry taht I did a mess😳. So to clarify for me. I have to merge the main branch locally into my branch and then squash my commits?

Charles-Gagnon commented:
You don't have to squash your commits - that'll happen when we eventually merge into main but it's fine to have a bunch of commits yourself. 

You do need to merge from main though otherwise Github gets confused and thinks that YOU made all the commits from main yourself so they show up as changes. 

Github provides a button to do this for you - https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork

Charles-Gagnon commented on Dockerfile:
```diff
@@ -0,0 +1,21 @@
+#This is a docker file to generate a SQL server. It helps you to execute the integration tests and for development
```
Let's do this in a separate PR so we can quickly review it there

M-Patrone replied:
Sure, that makes perfect sense

M-Patrone commented:
@Charles-Gagnon Thank you very much. I hope I did the right stuff 😉

Charles-Gagnon commented on src/Microsoft.SqlTools.ServiceLayer/Rename/RenameUtils.cs:
```diff
@@ -0,0 +1,74 @@
+            }
+            if (!Regex.IsMatch(requestParams.TableInfo.Schema, RegexMatchPatternValidObjectName) || !Regex.IsMatch(requestParams.TableInfo.Database, RegexMatchPatternValidObjectName) || !Regex.IsMatch(requestParams.TableInfo.OldName, RegexMatchPatternValidObjectName) || !Regex.IsMatch(requestParams.TableInfo.TableName, RegexMatchPatternValidObjectName) || !Regex.IsMatch(requestParams.ChangeInfo.N ...
+            {
+                throw new ArgumentOutOfRangeException(SR.NotAllowedCharacterForRenaming);
```
We should just let SMO/the engine do all the logic for this I think, not try to re-implement any of that ourselves.

M-Patrone replied:
Surely I can perfectly remove it 😊

Charles-Gagnon replied:
I'd remove the whole method

Charles-Gagnon commented on src/Microsoft.SqlTools.ServiceLayer/Rename/RenameService.cs:
```diff
@@ -0,0 +1,138 @@
+    /// <summary>
+    /// Main class for Rename Service functionality
+    /// </summary>
+    public class RenameService : IDisposable
```
@alanrenmsft I'm trying to decide here if we should make a more generic service that is used for generic modifications to objects, instead of a specialized service just for a single function. Thoughts on that? 

(and ideally making it something that other providers could implement as well)

alanrenmsft replied:
I thought about it, we probably should call it `ObjectManagementProvider` in dataprotocol  and in  sqltoolsservice, we can call the service `ObjectManagemetnService`), it is ok to start with only the rename functionality.  other features like create, retrieval, update and delete will be added later, this is something we are planning to do in the new semester.

Charles-Gagnon replied:
Yeah, something like that long term sounds great for sure. I figure we can get this logic in as just a single-purpose service for just MSSQL for now and then later on when/if we make it generic in the data protocol or whatever we can move the logic over there and fully implement that.

M-Patrone replied:
@Charles-Gagnon and @alanrenmsft is it ok, if I rename the ```RenameService``` to ```ObjectManagementService```?

Charles-Gagnon replied:
Yeah, that sounds great to me. And then the request ID would be `objectmanagement/rename` as Alan suggested

shueybubbles commented on src/Microsoft.SqlTools.ServiceLayer/Rename/RenameService.cs:
```diff
@@ -0,0 +1,138 @@
+               catch (Exception e)
+               {
+                   Logger.Error("Error on executing renaming operation: " + e.ToString());
+                   throw new InvalidOperationException("The renaming operation was not successfull executed");
```
shouldn't the exception message be a localizable resource?

M-Patrone replied:
That's absolutely true. I will add it.

alanrenmsft replied:
not sure why you picked InvalidOperationException type, that doesn't seem to be the right type, I don't mind if you use Exception directly.

shueybubbles replied:
I suggest that each component define its own exception type so the source of errors is very clear to developers and eventually to customer support.
Also it becomes very hard to debug something if you have to set your debugger exception filter to catch ALL first chance exceptions because your component throws "Exception". 
Choosing the scope for "Component" is perhaps interesting. If the tools service has already defined any common exception type it could be used here. Otherwise just creating a new type for this specific service and throw it, wrapping the handled exception as the InnerException.

shueybubbles commented on src/Microsoft.SqlTools.ServiceLayer/Rename/RenameService.cs:
```diff
@@ -0,0 +1,138 @@
+               {
+                   connectionService.TryFindConnection(
+                          requestParams.TableInfo.OwnerUri,
+                          out connInfo);
```
check the return value of TryFindConnection and return a useful message instead of trying to open a connection with a null connInfo

M-Patrone replied:
thats seems reasonable. I will add it.

shueybubbles commented on src/Microsoft.SqlTools.ServiceLayer/Rename/Requests/RenameTableInfo.cs:
```diff
@@ -0,0 +1,18 @@
+    public class RenameTableInfo
+    {
+        public string TableName { get; set; }
+        public string OldName { get; set; }
```
does the front end not have the Urn of the SMO object already? If not, you should consider finding out how to get it sent there when object explorer builds the tree. Otherwise this service is going to have so many special cases for all the object types it's going to get way too complicated.

M-Patrone replied:
I think we can't assume that every frontend has the urn of the object already. It is true, of course, that with this approach we have many cases to support (for the alpha version only tables and columns). If we would assume that not every frontend has an object explorer, this would have to be built in the frontend. What do you think about that?

shueybubbles replied:
I'm not versed enough on ADS architecture to know what the different frontends are that might use this service.

Charles-Gagnon replied:
It's true that not every provider will necessarily use URNs to identify their objects - but instead we can just assume that every provider will have a unique identifier (in our case the URN) to determine which object to update. 

So my suggestion here is to make the request this : 

```csharp
    public class RenameRequestParams
    {
        public string ObjectId { get; set; }
        public string NewName { get; set; }
    }

    public class RenameObjectRequest
    {
        public static readonly RequestType<RenameRequestParams, bool> Type = RequestType<RenameRequestParams, bool>.Create("rename/rename");
    }
```

And then for our case we'd assume (require) that ObjectId is the URN of the object, which we'd generate on the ADS side.  This will make it easily extensible for other providers since they can define what the object ID is themselves. 

Although I may also be overthinking this, if a provider wanted to implement their own rename they can just add a menu item contribution to OE for their own provider items....that would definitely make it simpler to design this since going and adding a rename feature to the core data protocol stuff would be a good chunk more work....

Let me think about it a bit and get back to you.

M-Patrone replied:
I think here also of applications outside of the ADS. At the moment the method ```GetSmoObject()``` only works via the Urn. If we now change to the "ObjectId" field, we need to specify that we need the Urn to be transmitted. This means especially precise specification work, especially in case someone wants to use this project outside of ADS. So in the end we need the Urn to use the method ```GetSmoObject()```. To sum up the field ``` ObjectId``` must be the Urn in this case.
If I think from the point of view of a developer, who wants to use this library in his own project. It is not clear that the ObjectId for that service is the Urn. So for an outside developer in my opinion, it is more clear to transmit the sql object name, which I want to be renamed. Maybe in the future it makes sense to write a service method that returns the Urn for an object name. This could be used if in a second step the rename service would be replaced by an “object change service”. For me the Urn is the ObjectId, which is used in this project or the "[sqlmanagementobjects](https://github.com/microsoft/sqlmanagementobjects)" project, so it is more an identification for intern purposes.
I am not sure, if I could explain myself. If not I am sorry 😉.

shueybubbles replied:
I think Charles means that any frontend has to get a reference to a SQL schema object from the tools service, whether it's from the same SQL provider that provides this rename service or some other provider that implements a rename service for its objects.
That object reference has to include an object id that can be used in these round trips and is opaque to the client. You wouldn't pass an objectid from a different provider to the rename service in this provider.

M-Patrone replied:
Yeah I thought of that only to use the service "rename". It is like only calling the service without the hassle to get the object reference like a "simple method call".

Charles-Gagnon replied:
What @shueybubbles said is correct, I was just trying to think in more general terms. 

Anyways after thinking about it I think it's best just to move forward with keeping this specific to MSSQL for now. If this is a feature that we want to design in a general way later on then we'll want to be much more deliberate and plan out what that would look like - which is beyond the scope of what I think is reasonable to do at this time. We can always design that feature later and then remove this stuff at that point if needed. 

@alanrenmsft FYI - we can chat offline if you have different thoughts

shueybubbles commented on src/Microsoft.SqlTools.ServiceLayer/Rename/RenameUtils.cs:
```diff
@@ -0,0 +1,74 @@
+
+            if (requestParams.ChangeInfo.Type == ChangeType.COLUMN)
+            {
+                return String.Format(urnTableReplaced + "/Column[@Name='{0}']", requestParams.TableInfo.OldName);
```
>Column[@Name='{0}']", requestParams.TableI [](http://example.com/codeflow?start=58&length=42)

if you are building Urns manually you need to urn escape the the parameters.

M-Patrone replied:
Ohh i did not think of that. Something like that?
```C# 
return String.Format(urnTableReplaced + @"/Column[@Name='{0}']", requestParams.TableInfo.OldName.ToString());
```
Do you think that is enough or should I add something specific?

shueybubbles replied:
look at `Urn.EscapeString`

M-Patrone replied:
Perfectly, thanks I will add it 👍

shueybubbles commented on src/Microsoft.SqlTools.ServiceLayer/Rename/RenameUtils.cs:
```diff
@@ -0,0 +1,74 @@
+            Server server = new Server(serverConnection);
+
+            Database database = new Database(server, requestParams.TableInfo.Database);
+            SqlSmoObject dbObject = database.Parent.GetSmoObject(new Urn(GetURNFromDatabaseSqlObjects(requestParams, serverConnection.TrueName)));
```
>etSmoOb [](http://example.com/codeflow?start=53&length=7)

i think GetSmoObject will work if the Urn doesn't have a Name attribute on the Server node.

M-Patrone replied:
Ok, I will remove it and check it.

M-Patrone replied:
So I removed the server name Name attribute. If I remove that, I get the following error:
```
Microsoft.SqlServer.Management.Smo.FailedOperationException: Attempt to retrieve data for object failed for Server '127.0.0.1,1433'
Microsoft.SqlServer.Management.Smo.SmoException: For the moment we don't support Urn's that do not start with Server
``` 
The Urn looks fine for me though: ```"Database[@Name='RenameTest_044efd534acc44e489ee03fd00712b42']/Table[@Name='testTable1_RenamingTable' and @Schema='dbo']"```

shueybubbles replied:
I meant you could just do "Server/Database[@Name=" instead of "Server[@Name='s']/Database" The Server node in the Urn is still required, just not the Name filter. Unless that's what you tried; your comment implies your Urn starts with Database though.

Charles-Gagnon commented on src/Microsoft.SqlTools.ServiceLayer/Rename/RenameService.cs:
```diff
@@ -0,0 +1,138 @@
+            return this.HandleRequest<bool>(requestContext, async () =>
+           {
+               Logger.Verbose("Handle Request in ProcessRenameEditRequest()");
+               bool operationExecutedSuccessFull = false;
```
Remove this - we only ever send the result if it succeeded so you can just hardcode it to `SendResult(true)` below

Charles-Gagnon commented on src/Microsoft.SqlTools.ServiceLayer/Rename/RenameService.cs:
```diff
@@ -0,0 +1,138 @@
+            {
+                renameObject.Rename(requestParams.ChangeInfo.NewName);
+            }
+            catch (Exception e)
```
Just remove this - we already catch it above and add additional info to the message so we can let this bubble up.

Charles-Gagnon commented on src/Microsoft.SqlTools.ServiceLayer/Rename/RenameService.cs:
```diff
@@ -0,0 +1,138 @@
+               catch (Exception e)
+               {
+                   Logger.Error("Error on executing renaming operation: " + e.ToString());
+                   throw new InvalidOperationException("The renaming operation was not successfull executed");
```
`throw new InvalidOperationException($"Error renaming {ObjectName}. {e.Message}, e)`

Charles-Gagnon commented on src/Microsoft.SqlTools.ServiceLayer/Rename/RenameUtils.cs:
```diff
@@ -0,0 +1,74 @@
+        {
+            String urnTableReplaced = String.Format(RenameUtils.UrnPrefix, trueNameOfServer, requestParams.TableInfo.Database, requestParams.TableInfo.TableName, requestParams.TableInfo.Schema, requestParams.TableInfo.OldName);
+
+            if (requestParams.ChangeInfo.Type == ChangeType.COLUMN)
```
Instead of having to pass the type and build up the URN here I think it would be better just to do this on the ADS side, similar to what I'm doing for the properties window : https://github.com/microsoft/azuredatastudio/blob/main/extensions/admin-tool-ext-win/src/utils.ts#L129

(and what we could look into to make this even easier is to actually pass the URN into the tree node from STS in the first place. Then we wouldn't even have to manually construct it ourselves - and the URN would be useful for a lot of other stuff too. But that can be done separately from this PR - for now I think it'd be best to just assume that we'll be getting the URN from ADS already fully constructed)

M-Patrone replied:
Thank you. I think you are right. Furthermore, I will change the properties in the class ```RenameRequestParams``` to that one:

```C#
public class RenameRequestParams
    {
        public string UrnOfObject{ get; set; }
        public string NewName { get; set; }
    }
```
I think that is more appropriate.

Charles-Gagnon commented:
@caohai Adding you here since Alan said you were talking to him about this feature earlier

alanrenmsft commented on src/Microsoft.SqlTools.ServiceLayer/Rename/RenameService.cs:
```diff
@@ -0,0 +1,138 @@
+    /// </summary>
+    public class RenameService : IDisposable
+    {
+        private bool disposed = false;
```
this doesn't seem to be used at all.

M-Patrone replied:
Of course, sorry it was a remnant from the old code.

alanrenmsft commented on src/Microsoft.SqlTools.ServiceLayer/Rename/RenameService.cs:
```diff
@@ -0,0 +1,138 @@
+        private IProtocolEndpoint serviceHost;
+
+        public RenameService() { }
+        /// <summary>
```
line break before this.

alanrenmsft commented on src/Microsoft.SqlTools.ServiceLayer/Rename/RenameService.cs:
```diff
@@ -0,0 +1,138 @@
+        }
+        internal Task HandleRequest<T>(RequestContext<T> requestContext, Func<Task> action)
+        {
+            // The request handling will take some time to return, we need to use a separate task to run the request handler so that it won't block the main thread.
```
looks like you copied the code from table designer :)

M-Patrone replied:
That's where i got the "inspiration" of how the projects works 😉

alanrenmsft commented on src/Microsoft.SqlTools.ServiceLayer/Rename/RenameService.cs:
```diff
@@ -0,0 +1,138 @@
+        {
+            return this.HandleRequest<bool>(requestContext, async () =>
+           {
+               Logger.Verbose("Handle Request in ProcessRenameEditRequest()");
```
the method name in the message doesn't match the actual name.

alanrenmsft commented on src/Microsoft.SqlTools.ServiceLayer/Rename/RenameService.cs:
```diff
@@ -0,0 +1,138 @@
+               }
+               catch (Exception e)
+               {
+                   Logger.Error("Error on executing renaming operation: " + e.ToString());
```
`An error occurred while renaming: `

alanrenmsft commented on src/Microsoft.SqlTools.ServiceLayer/Rename/RenameService.cs:
```diff
@@ -0,0 +1,138 @@
+        /// <param name="sqlConn">the sqlconnection on which the operation is executed</param>
+        private void ExecuteRenaming(ProcessRenameEditRequestParams requestParams, SqlConnection sqlConn)
+        {
+            Logger.Verbose("Inside in the ExecuteRenaming()-Method");
```
not sure whether we need this level of verbose logging, anyways, it probably should be `Entering the ExecuteRenaming method.`

Charles-Gagnon replied:
I'd just remove these, we generally aren't doing that anywhere else and if an exception occurs it should be clear where it came from anyways (verbose logging will include message information so it'll be clear it's from the rename)

M-Patrone replied:
I let myself be guided too firmly by school theory 😊. However, I can remove it with pleasure.

alanrenmsft commented on src/Microsoft.SqlTools.ServiceLayer/Rename/RenameService.cs:
```diff
@@ -0,0 +1,138 @@
+        /// </summary>
+        /// <param name="requestParams">parameters which are needed to execute renaming operation</param>
+        /// <param name="sqlConn">the sqlconnection on which the operation is executed</param>
+        private void ExecuteRenaming(ProcessRenameEditRequestParams requestParams, SqlConnection sqlConn)
```
do we really need this extra layer, you can move the 2 lines of the actual code to the `HandleProcessRenameEditRequest` method

alanrenmsft commented on src/Microsoft.SqlTools.ServiceLayer/Rename/Requests/ProcessRenameEditRequest.cs:
```diff
@@ -0,0 +1,22 @@
+    /// <summary>
+    /// Property class for Rename Service
+    /// </summary>
+    public class ProcessRenameEditRequestParams : GeneralRequestDetails
```
`RenameRequestParams`

alanrenmsft commented on src/Microsoft.SqlTools.ServiceLayer/Rename/Requests/ProcessRenameEditRequest.cs:
```diff
@@ -0,0 +1,22 @@
+        public RenameTableInfo TableInfo { get; set; }
+        public RenameTableChangeInfo ChangeInfo { get; set; }
+    }
+    public class ProcessRenameEditRequest
```
`RenameRequest`

alanrenmsft commented on src/Microsoft.SqlTools.ServiceLayer/Rename/Requests/ProcessRenameEditRequest.cs:
```diff
@@ -0,0 +1,22 @@
+    /// </summary>
+    public class ProcessRenameEditRequestParams : GeneralRequestDetails
+    {
+        public RenameTableInfo TableInfo { get; set; }
```
let's make this generic enough: `ObjectPath` or something similar that you can pass in from azure data studio side .

alanrenmsft replied:
and of course, to begin with, you can choose to only implement the table and column, but the whole thing should be very easy to extend.

M-Patrone replied:
That is absolutely true. As @Charles-Gagnon [pointed out](https://github.com/microsoft/sqltoolsservice/pull/1686#discussion_r972304258), I change the whole class to get the Urn of the object and the new name. It makes the whole operation much simpler.

alanrenmsft commented on src/Microsoft.SqlTools.ServiceLayer/Rename/Requests/ProcessRenameEditRequest.cs:
```diff
@@ -0,0 +1,22 @@
+    public class ProcessRenameEditRequestParams : GeneralRequestDetails
+    {
+        public RenameTableInfo TableInfo { get; set; }
+        public RenameTableChangeInfo ChangeInfo { get; set; }
```
`public String NewName { get; set; }`

alanrenmsft commented on src/Microsoft.SqlTools.ServiceLayer/Rename/Requests/ProcessRenameEditRequest.cs:
```diff
@@ -0,0 +1,22 @@
+    }
+    public class ProcessRenameEditRequest
+    {
+        public static readonly RequestType<ProcessRenameEditRequestParams, bool> Type = RequestType<ProcessRenameEditRequestParams, bool>.Create("rename/processedit");
```
the type string should be: `objectmanagement/rename`

M-Patrone commented:
@All Thank you very much for the effort and time to get me feedback. I really appreciate the detailed codereview. Thus, I had some improvments to do.  Here a short list of all the changes:

- I removed the helper class RenameUtils 
- Rename the RenameService to ObjectManagementService  
- Change the InvalidOperationException to Exception. I also changed the message to the recommendation from @alanrenmsft
- Surround TryFindConnection with an if. Now it throws error when connection is not found 
- Replace variable operationExecutedSuccessFull 
- Rename “ProcessRenameEditRequestParams” to “RenameRequestParams” 
- Remove the IDisposible Interface with the associated methods 
- Line break after constructor 
- Logger: change method name to “HandleProcessRenameEditRequest” 
- Remove the Method ExecuteRenaming() 
- Rename “ProcessRenameEditRequest” to “RenameRequest”  
- Remove escape because of other services 
- Add integration test for connection not found

If I have to change something, let me know.

shueybubbles commented on src/Microsoft.SqlTools.ServiceLayer/ObjectManagement/ObjectManagementService.cs:
```diff
@@ -0,0 +1,126 @@
+                   else
+                   {
+                       Logger.Error("The connection could not be found.");
+                       throw new Exception(SR.ErrorConnectionNotFound);
```
> throw new Exception(SR.ErrorConnectionNotFound); [](http://example.com/codeflow?start=22&length=49)

why is throwing a plain "Exception" considered a good practice? 
Typically, components define their own exception types so they are isolatable from other component exceptions during debugging.

M-Patrone replied:
Initially I would agree with you. But I interpreted [this comment](https://github.com/microsoft/sqltoolsservice/pull/1686#discussion_r972291780), that I should use the ```Exception``` class. Should I create a separate exception or not?

Charles-Gagnon replied:
It's just not something we've ever really done here - @shueybubbles's points are valid but for now let's keep it with a plain Exception and I'll make a note to follow up and make some sort of general "ToolsServiceException" we can use instead of plain Exception objects so that we can update everything at once.

shueybubbles commented on src/Microsoft.SqlTools.ServiceLayer/ObjectManagement/ObjectManagementService.cs:
```diff
@@ -0,0 +1,126 @@
+               catch (Exception e)
+               {
+                   Logger.Error($"An error occurred while renaming {requestParams.UrnOfObject}: " + e);
+                   throw new Exception(SR.ErrorOnRenameOperationMessage, e);
```
what's the value of catching your own exception and rethrowing it as an inner?

M-Patrone replied:
Here I oriented me on [this suggestion](https://github.com/microsoft/sqltoolsservice/pull/1686#discussion_r972225476). For me the point was that I get to know where the error happened. The inner exception is to know the “cause” of the error ([suggestion](https://github.com/microsoft/sqltoolsservice/pull/1686#discussion_r972224684)). Also, I thought of developer perspective it adds additional information to an exception, which is not interesting for a user.

Charles-Gagnon replied:
Thinking about it we can just remove this to simplify it - if an error occurs then ADS will know it was trying to rename something so can add additional messaging as needed.

M-Patrone replied:
@Charles-Gagnon do I understand that correctly, that I should remove the Innerexception or do you mean the whole try-catch?

Charles-Gagnon replied:
The whole try catch, the dispatcher will already log and send the error itself if one occurs.

shueybubbles commented on src/Microsoft.SqlTools.ServiceLayer/ObjectManagement/ObjectManagementService.cs:
```diff
@@ -0,0 +1,126 @@
+        {
+            ServerConnection serverConnection = new ServerConnection(connection);
+            Server server = new Server(serverConnection);
+            SqlSmoObject dbObject = server.GetSmoObject(new Urn(requestParams.UrnOfObject));
```
>GetSmoO [](http://example.com/codeflow?start=43&length=7)

does this throw if it can't find the object, or does it return null? if it returns null, you should check for it in the caller and return a specific error about the object not being found. Otherwise the exception message is going to be for a null reference

Charles-Gagnon replied:
Looks like it throws 

https://github.com/microsoft/sqlmanagementobjects/blob/main/src/Microsoft/SqlServer/Management/Smo/serverbase.cs#L668

M-Patrone replied:
I do not think, that null will be returned (As i understand it, it throws an [MissingObjectException](https://github.com/microsoft/sqlmanagementobjects/blob/04249a2dc85ad1aa0d492b8a2cecf65583ded177/src/Microsoft/SqlServer/Management/Smo/serverbase.cs#L666)). I also tested for Object, which [do not exist](https://github.com/M-Patrone/sqltoolsservice/blob/feature-rename-sql-objects/test/Microsoft.SqlTools.ServiceLayer.IntegrationTests/ObjectManagement/ObjectManagementServiceTests.cs#L82). Should I add it nevertheless?

shueybubbles replied:
since SMO throws, no need to check for null. thx

alanrenmsft commented on src/Microsoft.SqlTools.ServiceLayer/Localization/sr.strings:
```diff
@@ -2415,4 +2415,10 @@ TableDesignerConfirmationText = I have read the summary and understand the poten
+#ObjectManagement Service
+NotAllowedCharacterForRenaming = The name contains characters which are not allowed
+RenameRequestParametersNotNullOrEmpty = The arguments must not be null nor empty
+ErrorOnRenameOperationMessage = The renaming operation was not successfull executed
```
`Error occurred while executing the renaming operation`

alanrenmsft commented on src/Microsoft.SqlTools.ServiceLayer/ObjectManagement/ObjectManagementService.cs:
```diff
@@ -0,0 +1,126 @@
+        /// <param name="requestParams">parameters which are needed to execute renaming operation</param>
+        /// <param name="requestContext">Request Context</param>
+        /// <returns></returns>
+        internal Task HandleProcessRenameEditRequest(RenameRequestParams requestParams, RequestContext<bool> requestContext)
```
HandleRenameRequest

alanrenmsft commented on src/Microsoft.SqlTools.ServiceLayer/ObjectManagement/ObjectManagementService.cs:
```diff
@@ -0,0 +1,126 @@
+                           requestParams.OwnerUri,
+                           out connInfo))
+                   {
+                       using (SqlConnection sqlConn = ConnectionService.OpenSqlConnection(connInfo, "RenamingDatabaseObjects"))
```
I am actually not sure whether we should have different application names within sql tools service. since it is already the case, let's keep it consistent, please define a string constant in this class ` public const string ObjectManagementServiceApplicationName = "azdata-object-management";` and  use it here.

Charles-Gagnon replied:
So that it can be re-used for other places in this service later on if needed. You could probably make it private for now though since we don't actually have a use case for that yet.

alanrenmsft commented on src/Microsoft.SqlTools.ServiceLayer/ObjectManagement/Requests/ProcessRenameEditRequest.cs:
```diff
@@ -0,0 +1,23 @@
+    {
+        public string UrnOfObject { get; set; }
+        public string NewName { get; set; }
+        public string OwnerUri { get; set; }
```
let's call this `ConnectionUri`

alanrenmsft commented on src/Microsoft.SqlTools.ServiceLayer/ObjectManagement/Requests/ProcessRenameEditRequest.cs:
```diff
@@ -0,0 +1,23 @@
+    /// </summary>
+    public class RenameRequestParams : GeneralRequestDetails
+    {
+        public string UrnOfObject { get; set; }
```
`ObjectUrn`

Charles-Gagnon commented on src/Microsoft.SqlTools.ServiceLayer/Localization/sr.strings:
```diff
@@ -2415,4 +2415,10 @@ TableDesignerConfirmationText = I have read the summary and understand the poten
+GetUserDefinedObjectsFromModelFailed = Failed to get user defined objects from model.
+
+#ObjectManagement Service
+NotAllowedCharacterForRenaming = The name contains characters which are not allowed
```
You can remove these validation messages since you're not using them anymore.

M-Patrone replied:
Yes, I saw it at the moment as I pushed it ;)

Charles-Gagnon commented on src/Microsoft.SqlTools.ServiceLayer/ObjectManagement/ObjectManagementService.cs:
```diff
@@ -0,0 +1,126 @@
+            this.serviceHost.SetRequestHandler(RenameRequest.Type, HandleProcessRenameEditRequest);
+
+        }
+        internal Task HandleRequest<T>(RequestContext<T> requestContext, Func<Task> action)
```
Just remove this HandleRequest method and just have HandleProcessRenameEditRequest - this isn't adding anything useful (the message dispatcher will handle starting another thread if needed and it handles sending an error if any exception is thrown)

M-Patrone commented on src/Microsoft.SqlTools.ServiceLayer/ObjectManagement/ObjectManagementService.cs:
```diff
@@ -0,0 +1,126 @@
+        /// <param name="requestParams">parameters which are required for the rename operation</param>
+        /// <param name="connection">the sqlconnection on the server to search for the sqlobject</param>
+        /// <returns>the sql object if implements the interface IRenamable, so they can be renamed</returns>
+        private IRenamable GetSQLRenameObject(RenameRequestParams requestParams, SqlConnection connection)
```
I’m calling the from the using and the sqlconnection is passed as a reference. Should it not also dispose the connection if the end of the using is reached (that is a general question I really don’t know)? I took the ```AgentAction```-Class as an example.

I would still make this in a separate ethod, to separate and structure the code. What do you think of that approach?
```C#

        private IRenamable GetSQLRenameObject(RenameRequestParams requestParams, ServerConnection serverConnection)
        {
            if (!serverConnection.IsOpen)
            {
                serverConnection.Connect();
            }
            Server server = new Server(serverConnection);
            SqlSmoObject dbObject = server.GetSmoObject(new Urn(requestParams.ObjectUrn));


            return (IRenamable)dbObject;
        }
```
I would use the ServerConnection as a parameter. In the ```HandleRenameRequest``` I would then use the following code in the catch part:
```C#
if (serverConnection != null && serverConnection.IsOpen)
                   {
                       serverConnection.Disconnect();
                   }
```

Charles-Gagnon replied:
Neither of them are actually, and the SqlConnection is being disposed above so this should be fine. Could definitely merge it into the main method above though given how simple this is (and unlikely to be reused).

Charles-Gagnon replied:
To answer your question - yes disposing the SqlConnection is enough (that handles closing the connection which is the main thing we care about). 

Given that you open the connection in the method above what you have here is fine - no need to complicate it further with open checks. And if you want to keep it as a utility method that should be fine too.

M-Patrone replied:
Perfectly, thank you very much.

Charles-Gagnon commented on src/Microsoft.SqlTools.ServiceLayer/ObjectManagement/ObjectManagementService.cs:
```diff
@@ -0,0 +1,126 @@
+                   }
+                   else
+                   {
+                       Logger.Error("The connection could not be found.");
```
Include the URI in this log message as well

Charles-Gagnon commented on src/Microsoft.SqlTools.ServiceLayer/ObjectManagement/Requests/ProcessRenameEditRequest.cs:
```diff
@@ -0,0 +1,23 @@
+    /// <summary>
+    /// Property class for ObjectManagement Service
+    /// </summary>
+    public class RenameRequestParams : GeneralRequestDetails
```
1. Rename the `Requests` folder to `Contracts` (to match what we do elsewhere)
2. Rename this file to `RenameRequest.cs`

M-Patrone commented:
Sorry for the  constant requests, but I want to make it as good as possible. Therefore I made some imporvements to the code: 

- Rename the folder “Requests” to “Contracts” 
- Added constant “ObjectManagementServiceApplicationName” 
- Removed the Method “HandleRequest” 
- Rename the “HandleRenameRequest”-Method to “HandleRenameRequest”-Method
- Removed the try catch block in the “HandleRenameRequest”-Method.  
- Improve the Integrationtests 
- Remove of the error message “ErrorOnRenameOperationMessage” 
- Rename the method to “HandleRenameRequest” 
- Property change “OwnerUri” to “ConnectionUri” 
- Property change “UrnOfObject” to “ObjectUrn” 
- Remove of the “HandleRequest”-Method. @Charles-Gagnon did I understand [that](https://github.com/microsoft/sqltoolsservice/pull/1686#discussion_r973186407) correctly, that we do not need “requestContext.SendError();”?

Charles-Gagnon commented on src/Microsoft.SqlTools.ServiceLayer/ObjectManagement/ObjectManagementService.cs:
```diff
@@ -0,0 +1,104 @@
+        {
+            get
+            {
+                return ConnectionService.Instance;
```
```
connectionService ??= ConnectionService.Instance;
return connectionService;
```

Charles-Gagnon commented on src/Microsoft.SqlTools.ServiceLayer/ObjectManagement/Contracts/RenameRequest.cs:
```diff
@@ -0,0 +1,23 @@
+    /// </summary>
+    public class RenameRequestParams : GeneralRequestDetails
+    {
+        public string ObjectUrn { get; set; }
```
Add doc comments for each of these

Charles-Gagnon commented on src/Microsoft.SqlTools.ServiceLayer/ObjectManagement/Contracts/RenameRequest.cs:
```diff
@@ -0,0 +1,23 @@
+namespace Microsoft.SqlTools.ServiceLayer.ObjectManagement.Contracts
+{
+    /// <summary>
+    /// Property class for ObjectManagement Service
```
Remove - it's incorrect and generally not useful enough to bother adding doc comments for (name is clear enough)

Charles-Gagnon commented on src/Microsoft.SqlTools.ServiceLayer/ObjectManagement/Contracts/RenameRequest.cs:
```diff
@@ -0,0 +1,29 @@
+    public class RenameRequestParams : GeneralRequestDetails
+    {
+        /// <summary>
+        /// uniform resource name of the object  
```
`SFC (SMO) URN identifying the object`

M-Patrone replied:
I will replace my comment

Charles-Gagnon commented on test/Microsoft.SqlTools.ServiceLayer.IntegrationTests/ObjectManagement/ObjectManagementServiceTests.cs:
```diff
@@ -0,0 +1,171 @@
+            requestContextMock.Verify(x => x.SendResult(It.Is<bool>(r => VerifySendedFeedback(r, true))));
+            await CheckGeneratedScript(testDb.DatabaseName, "CREATE TABLE [dbo].[RenamingTable]");
+
+            await SqlTestDb.DropDatabase(testDb.DatabaseName);
```
The drop should be done as part of a teardown function so that a failure in the test doesn't cause it to skip cleanup

Charles-Gagnon commented on test/Microsoft.SqlTools.ServiceLayer.IntegrationTests/ObjectManagement/ObjectManagementServiceTests.cs:
```diff
@@ -0,0 +1,171 @@
+        public async Task TestRenameTable()
+        {
+            await objectManagementService.HandleRenameRequest(this.InitRequestParams("RenamingTable", String.Format("Server/Database[@Name='{0}']/Table[@Name='testTable1_RenamingTable' and @Schema='dbo']", testDb.DatabaseName)), requestContextMock.Object);
+            Thread.Sleep(2000);
```
Why the sleeps?

M-Patrone replied:
Sorry, that was a remnant of my old laptop 😉 Of course I will remove it!

Charles-Gagnon commented on test/Microsoft.SqlTools.ServiceLayer.IntegrationTests/ObjectManagement/ObjectManagementServiceTests.cs:
```diff
@@ -0,0 +1,171 @@
+            await objectManagementService.HandleRenameRequest(this.InitRequestParams("RenamingTable", String.Format("Server/Database[@Name='{0}']/Table[@Name='testTable1_RenamingTable' and @Schema='dbo']", testDb.DatabaseName)), requestContextMock.Object);
+            Thread.Sleep(2000);
+            requestContextMock.Verify(x => x.SendResult(It.Is<bool>(r => VerifySendedFeedback(r, true))));
+            await CheckGeneratedScript(testDb.DatabaseName, "CREATE TABLE [dbo].[RenamingTable]");
```
This seems more complex than I think it needs to be - you can just run the rename and verify that the old table name doesn't exist but the new one does with a direct query.

M-Patrone replied:
Sorry, only one query. Do I understand that correctly, that I should execute a SQL-Query with the [OBJECT_ID-Function](https://learn.microsoft.com/en-us/sql/t-sql/functions/object-id-transact-sql?view=sql-server-ver16) from TSQL to prove if the item exists or even not?

Charles-Gagnon replied:
Something like that - yeah. (even just checking sys.tables should be enough but the general concept is the same)

Charles-Gagnon commented on test/Microsoft.SqlTools.ServiceLayer.IntegrationTests/ObjectManagement/ObjectManagementServiceTests.cs:
```diff
@@ -0,0 +1,171 @@
+            return result == expectedResult;
+        }
+
+        private RenameRequestParams InitRequestParams(string newName, string UrnOfObject)
```
Doesn't seem like something that needs a helper function given that all it's doing is creating a single object

M-Patrone replied:
I am not sure. Is it not redudant if I have to create the object in every test-method? For me personally I would say that it's more intuitive and saves also a lot of code lines.

Charles-Gagnon commented on test/Microsoft.SqlTools.ServiceLayer.IntegrationTests/ObjectManagement/ObjectManagementServiceTests.cs:
```diff
@@ -0,0 +1,171 @@
+
+        private static bool VerifySendedFeedback(bool result, bool expectedResult)
+        {
+            return result == expectedResult;
```
This doesn't seem necessary - just do the comparison directly in each test.

Charles-Gagnon commented on test/Microsoft.SqlTools.ServiceLayer.IntegrationTests/ObjectManagement/ObjectManagementServiceTests.cs:
```diff
@@ -0,0 +1,171 @@
+        [SetUp]
+        public async Task TestInitialize()
+        {
+            this.testDb = await SqlTestDb.CreateNewAsync(TestServerType.OnPrem, false, null, table_query, "RenameTest");
```
Instead of passing null for the values you don't care about I'd suggest using named parameters for each of the other ones (which also benefits the reader to know what the values mean without having to dig into the method signature)

M-Patrone replied:
Thanks, I will remove it and I will also remove the parameters which would have the default value specified on the method.

Charles-Gagnon commented on test/Microsoft.SqlTools.ServiceLayer.IntegrationTests/ObjectManagement/ObjectManagementServiceTests.cs:
```diff
@@ -0,0 +1,171 @@
+    public class ObjectManagementServiceTests
+    {
+        private readonly string table_query = @"CREATE TABLE testTable1_RenamingTable (c1 int)
+                            GO
```
Is the GO actually necessary? I'd remove it if not so the query is on a single line

Charles-Gagnon commented on test/Microsoft.SqlTools.ServiceLayer.IntegrationTests/ObjectManagement/ObjectManagementServiceTests.cs:
```diff
@@ -0,0 +1,171 @@
+{
+    public class ObjectManagementServiceTests
+    {
+        private readonly string table_query = @"CREATE TABLE testTable1_RenamingTable (c1 int)
```
This can be a constant

Charles-Gagnon commented on test/Microsoft.SqlTools.ServiceLayer.IntegrationTests/ObjectManagement/ObjectManagementServiceTests.cs:
```diff
@@ -0,0 +1,171 @@
+                            GO
+                            ";
+        private readonly string ownerUri = "testDB";
+        private readonly string startTableName = "testTable1_RenamingTable";
```
These start* values don't seem to be used

Charles-Gagnon commented on test/Microsoft.SqlTools.ServiceLayer.IntegrationTests/ObjectManagement/ObjectManagementServiceTests.cs:
```diff
@@ -0,0 +1,171 @@
+            return true;
+        }
+
+        protected void VerifyErrorSent<T>(Mock<RequestContext<T>> contextMock)
```
This doesn't seem to be used

Charles-Gagnon reviewed (changes requested):
The functionality seems fine, just some cleanup with the tests!

M-Patrone commented:
Please do not do the review. I am not finished yet 😊

M-Patrone commented:
Sorry for my late response. I have now made the following optimizations:
- change <summarize> comment
- adding teardown-method
- removing the sleep
- check tests with sys.tables
- named parameters in the TestInitialize()-method
- Removing the "GO"
- adding constant
- remove unnecessary properties
- remove error verification

The only thing I didn't do is remove the method to create my test object.

Charles-Gagnon commented on test/Microsoft.SqlTools.ServiceLayer.IntegrationTests/ObjectManagement/ObjectManagementServiceTests.cs:
```diff
@@ -0,0 +1,150 @@
+            Query queryRenameObject = ExecuteQuery("SELECT * FROM " + testDb.DatabaseName + ".sys.tables WHERE name='RenamingTable'");
+            Assert.IsTrue(queryRenameObject.HasExecuted);
+            Assert.IsFalse(queryRenameObject.HasErrored);
+            Assert.AreEqual(1, queryRenameObject.Batches[0].ResultSets[0].RowCount);
```
For these asserts it's preferred to use the fluent (constraint) style assertions (https://docs.nunit.org/articles/nunit/writing-tests/assertions/assertion-models/constraint.html) - e.g. this would be 

`Assert.That(queryRenameObject.Batches[0].ResultSets[0].RowCount, Is.EqualTo(1))`

They're a lot easier to read, are harder to mess up order of expected vs actual and have better error messages when they do fail. (all assertions in these tests can be changed to use `Assert.That(...)`)

In addition when you do this make sure you add a message to each of them that indicates what failed, since otherwise you'll usually just end up with a message like "0 does not equal 1" which isn't helpful, especially when there's multiple similar assertions in the same test. 

e.g. 

`Assert.That(queryRenameObject.Batches[0].ResultSets[0].RowCount, Is.EqualTo(1), "Did not find table with new name after rename")`

M-Patrone replied:
Yeah sure, I can do that. Thank you very much for the tipp. Are you ok, with the error messages (not a native english speaker! 😉)?

```c#
 Query queryRenameObject = ExecuteQuery("SELECT * FROM " + testDb.DatabaseName + ".sys.tables WHERE name='RenamingTable'");
            Assert.That(queryRenameObject.HasExecuted, Is.True, "The query to check for the renamed table was not executed");
            Assert.That(queryRenameObject.HasErrored, Is.False, "There were errors on the execution of the query to check for the renamed table");
            Assert.That(queryRenameObject.Batches[0].ResultSets[0].RowCount, Is.EqualTo(1), "Did not find the table with the new name after the rename operation");

            Query queryOldObject = ExecuteQuery("SELECT * FROM " + testDb.DatabaseName + ".sys.tables WHERE name='testTable1_RenamingTable'");
            Assert.That(queryOldObject.HasExecuted, Is.True, "The query to check for the old table was not executed");
            Assert.That(queryOldObject.HasErrored, Is.False, "There were errors on the execution of the query to check for the old table");
            Assert.That(queryOldObject.Batches[0].ResultSets[0].RowCount, Is.EqualTo(0), "Did find the old table which should have been renamed");
        
```

Charles-Gagnon replied:
Yeah, those are fine! It's mostly just to have a unique phrase so it's clear where exactly the failure happened. 

(and do this for the rest of the tests in this file too)

Charles-Gagnon commented on test/Microsoft.SqlTools.ServiceLayer.IntegrationTests/ObjectManagement/ObjectManagementServiceTests.cs:
```diff
@@ -0,0 +1,150 @@
+    public class ObjectManagementServiceTests
+    {
+        private const string TableQuery = @"CREATE TABLE testTable1_RenamingTable (c1 int)";
+        private const string OWNERURI = "testDB";
```
`OwnerUri`

Charles-Gagnon reviewed (approved):
@shueybubbles Anything else blocking on your end?

@alanrenmsft Any last comments?

Looks good to me though once you fix up the assertions. Do you know what you need to do on the ADS side to get this stuff hooked up?

alanrenmsft commented on src/Microsoft.SqlTools.ServiceLayer/ObjectManagement/ObjectManagementService.cs:
```diff
@@ -0,0 +1,105 @@
+        /// <param name="requestParams">parameters which are required for the rename operation</param>
+        /// <param name="connection">the sqlconnection on the server to search for the sqlobject</param>
+        /// <returns>the sql object if implements the interface IRenamable, so they can be renamed</returns>
+        private IRenamable GetSQLRenameObject(RenameRequestParams requestParams, SqlConnection connection)
```
`GetRenamable`

shueybubbles commented on test/Microsoft.SqlTools.ServiceLayer.Test.Common/SqlTestDb.cs:
```diff
@@ -25,7 +25,6 @@ public class SqlTestDb : IDisposable
 
         public bool DoNotCleanupDb { get; set; }
-
         public string ConnectionString
```
if this file is just whitespace changes you can revert

shueybubbles reviewed (approved):
:shipit:

alanrenmsft commented on src/Microsoft.SqlTools.ServiceLayer/ObjectManagement/ObjectManagementService.cs:
```diff
@@ -0,0 +1,105 @@
+                throw new Exception(SR.ErrorConnectionNotFound);
+            }
+
+            await requestContext.SendResult(true);
```
does it must have a value? this is never going to be false.

Charles-Gagnon replied:
Yes, limitation of our request context stuff currently. I want to go fix that at some point to allow an empty (void) response but for now this is fine.

alanrenmsft commented on src/Microsoft.SqlTools.ServiceLayer/ObjectManagement/ObjectManagementService.cs:
```diff
@@ -0,0 +1,105 @@
+            ServerConnection serverConnection = new ServerConnection(connection);
+            Server server = new Server(serverConnection);
+            SqlSmoObject dbObject = server.GetSmoObject(new Urn(requestParams.ObjectUrn));
+
```
remove the unnecessary line breaks

M-Patrone commented:
Do not review the PR yet. I have to execute the tests on another machine.

Charles-Gagnon commented:
Really appreciate all the work you've done here @M-Patrone - let us know if you need help on where to get started for getting this hooked up on the ADS side

(to start with - this will be useful to get ADS using a local copy of STS which has your changes in it https://github.com/microsoft/azuredatastudio/wiki/SQL-Tools-Service#using-locally-built-binaries-for-debuggingtesting)

M-Patrone commented:
> Really appreciate all the work you've done here @M-Patrone - let us know if you need help on where to get started for getting this hooked up on the ADS side
> 
> (to start with - this will be useful to get ADS using a local copy of STS which has your changes in it https://github.com/microsoft/azuredatastudio/wiki/SQL-Tools-Service#using-locally-built-binaries-for-debuggingtesting)

Special thanks to @Charles-Gagnon , @alanrenmsft and @shueybubbles ! I really appreciate the help and the honest code reviews, which you all gave me. I learned a lot. If I could do something for you let me know!

For me this and the ADS project is also to learn how to adapt to complex projects and how to add new features to them. I am really sorry that I wasted so much time of yours and I know that’s not the right projects to exercises or that’s your duty to take time to educate me on this. 

@Charles-Gagnon  thank you very much for the starting guide. With that I found out that I should add the logic to the sqltoolsservices project. 😊

You also asked me the question if I know where to start on this to implement this feature on ADS. To be honest, I don’t know 😟. But I would look after similar features to implement that one (like refresh, Object Explorer). I really hope that I don’t make so many errors/coding-guidelines errors this time.

Charles-Gagnon commented:
So for renaming it would be nice to hook into the F2 -> edit box functionality that the explorer (file) view has. But I looked into that and it seems like a custom thing for that view. We'll have to look more into that to see if that's something we can easily hook up for OE.

For now though I'd suggest using the "Select Top 1000" action as a guide, since that's very similar to what you're doing. This will get us context menu action items for the nodes that users can use to rename an object, which while not ideal is at least something to get started with.

There's two main parts - the action definition and the code to send the request to the server. 

For the action you'll start by creating an action class similar to [ScriptSelectAction](https://github.com/Microsoft/azuredatastudio/blob/3cb2f552a6f004dccd153577aecfd75f3d1bacae/src/sql/workbench/browser/scriptingActions.ts#L15).

Then you'll register it with the command registry in https://github.com/Microsoft/azuredatastudio/blob/3cb2f552a6f004dccd153577aecfd75f3d1bacae/src/sql/workbench/contrib/scripting/browser/scriptingActions.ts (note there will be two separate command IDs - one for OE and one for the Azure View, also called Data Explorer in this context)

The action is going to need to do a couple things, first is prompt the user for the new name (using [IQuickInputService.input(...)](https://github.com/Microsoft/azuredatastudio/blob/2bc6a0cd012de69504061a876070cfa43533fedc/src/vs/platform/quickinput/common/quickInput.ts#L53))

And let's just start with that - so once the above is done you should see the menu item and when selected have it prompt for a name (but not do anything beyond that). After that's done we can help walk through how to hook that up to send to STS.
